### PR TITLE
Wascc logs

### PR DESCRIPTION
**Pull request by bketelsen**
This PR adds the capability to store and retrieve actor logs from the wascc runtime.

### CODE REVIEW

File: crates/wascc-logging/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,150 @@
+// Copyright 2015-2019 Capital One Services, LLC
```
Comment (thomastaylor312) [inline crates/wascc-logging/src/lib.rs:1]: We should probably note here that this is adopted from the wascc logging crate with the original license included here

File: crates/wascc-logging/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+    INFO,
+    DEBUG,
+    TRACE,
+}
+
+/// LoggingProvider provides an implementation of the wascc:logging capability
+/// that keeps separate log output for each actor.
+pub struct LoggingProvider {
+    dispatcher: RwLock<Box<dyn Dispatcher>>,
+    output_map: RwLock<HashMap<String, Box<WriteLogger<File>>>>,
```
[the remark this answers is not in the document]
Comment (thomastaylor312) [reply crates/wascc-logging/src/lib.rs:56]: Didn't want to unbox it. The `new` function returns a box

File: crates/kubelet/src/handle.rs [hunk 3]
```diff
@@ -190,3 +190,15 @@ impl<R: AsyncRead + AsyncSeek + Unpin, S: Stop> PodHandle<R, S> {
         Ok(())
     }
 }
+
+/// Generates a unique human readable key for storing a handle to a pod in a
+/// hash. This is a convenience wrapper around [crate::pod_key].
```
Comment (rylev) [inline crates/kubelet/src/handle.rs:195]: `pod_key` isn't exposed at the top level of the crate. This should just be ```[`pod_key`]```

File: crates/wascc-logging/src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (rylev) [reply crates/wascc-logging/src/lib.rs:56]: Looking at the source of WriteLogger I'm also confused why it boxes. Are we sure that simplelog is at simple as it reports to be? ;-)

File: crates/kubelet/src/handle.rs [hunk 4]
```diff
@@ -190,3 +190,15 @@ impl<R: AsyncRead + AsyncSeek + Unpin, S: Stop> PodHandle<R, S> {
...
+pub fn key_from_pod(pod: &Pod) -> String {
+    pod_key(pod.namespace(), pod.name())
+}
+
+// Generates a unique human readable key for storing a handle to a pod if you
+// already have the namespace and pod name.
+pub fn pod_key<N: AsRef<str>, T: AsRef<str>>(namespace: N, pod_name: T) -> String {
```
Comment (rylev) [inline crates/kubelet/src/handle.rs:202]: Does this have to be public? Seems like `key_from_pod`  is enough

File: crates/wascc-logging/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+            println!("Received log call {}", actor);
+            let log_msg = deserialize::<WriteLogRequest>(msg)?;
+
+            println!("[Level:{}] {}", log_msg.level, log_msg.body);
+            let output_map = self.output_map.read().unwrap();
+            let logger = output_map
+                .get(actor)
+                .ok_or(format!("unable to find logger for actor {}", actor))?;
+            logger.log(
+                &log::Record::builder()
```
[the remark this answers is not in the document]
Comment (bketelsen) [reply crates/wascc-logging/src/lib.rs:124]: won't work.  it has to be inlined or there will be lifetime issues.  We started with it outside.

File: crates/kubelet/src/handle.rs (hunk 4 above)
Comment (bketelsen) [reply crates/kubelet/src/handle.rs:202]: yes, there are places we don't have a pod but do have a name

File: crates/wascc-logging/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+                    .args(format_args!("[{}] {}", actor, log_msg.body))
+                    .level(match log_msg.level {
```
[the remark this answers is not in the document]
Comment (bketelsen) [reply crates/wascc-logging/src/lib.rs:130]: I'll open an issue for this, it's harder than I thought :)

File: crates/kubelet/src/handle.rs (hunk 4 above)
Comment (thomastaylor312) [reply crates/kubelet/src/handle.rs:202]: Yeah, this is used for `logs` as it passes the namespace and name, but not the pod object

File: crates/wascc-logging/src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (thomastaylor312) [reply crates/wascc-logging/src/lib.rs:60]: I asked the same question. Don't know exactly why it is boxed

**Pull request merged**